Hello world session cannot be deserialized

A serialized session for hummingbird-demo/demo-hello-world (changed to use a serializable inner class instead of a lambda for the event listener) cannot be deserialized. This happens every time you restart a Tomcat server where the app is deployed.

The last part of the stacktrace is:

```
Caused by: java.lang.ClassCastException: java.lang.invoke.SerializedLambda cannot be cast to com.vaadin.hummingbird.dom.DomEventListener
        at com.vaadin.hummingbird.namespace.ElementListenersNamespace.$deserializeLambda$(ElementListenersNamespace.java:1)
```

--- a/hummingbird-server/src/main/java/com/vaadin/hummingbird/event/ComponentEventBus.java
+++ b/hummingbird-server/src/main/java/com/vaadin/hummingbird/event/ComponentEventBus.java
@@ -86,7 +86,14 @@ public class ComponentEventBus implements Serializable {
                 .computeIfAbsent(eventType,
                         t -> new ComponentEventData()).listeners;
         listeners.add(listener);
-        return () -> removeListener(eventType, listener);
+        // This needs to be an anonymous class and not a lambda because of
+        // https://github.com/vaadin/hummingbird/issues/575
+        return new EventRegistrationHandle() {
+            @Override
+            public void remove() {
+                removeListener(eventType, listener);
+            }
+        };
     }
 
     /**

--- a/hummingbird-tests/test-root-context/src/main/java/com/vaadin/hummingbird/uitest/servlet/ViewClassLocator.java
+++ b/hummingbird-tests/test-root-context/src/main/java/com/vaadin/hummingbird/uitest/servlet/ViewClassLocator.java
@@ -79,7 +79,7 @@ public class ViewClassLocator {
         return Logger.getLogger(ViewClassLocator.class.getName());
     }
 
-    Collection<Class<? extends View>> getAllViewClasses() {
+    public Collection<Class<? extends View>> getAllViewClasses() {
         return views.values();
     }
 

--- a/hummingbird-tests/test-root-context/src/main/java/com/vaadin/hummingbird/uitest/ui/ForTemplateView.java
+++ b/hummingbird-tests/test-root-context/src/main/java/com/vaadin/hummingbird/uitest/ui/ForTemplateView.java
@@ -25,6 +25,7 @@ import com.vaadin.hummingbird.nodefeature.ModelList;
 import com.vaadin.hummingbird.nodefeature.ModelMap;
 import com.vaadin.hummingbird.nodefeature.TemplateOverridesMap;
 import com.vaadin.hummingbird.router.View;
+import com.vaadin.server.Command;
 import com.vaadin.ui.Template;
 
 /**
@@ -70,9 +71,9 @@ public class ForTemplateView extends Div implements View {
                 () -> modelList.remove(0)));
     }
 
-    private Button createButton(String text, String id, Runnable action) {
+    private Button createButton(String text, String id, Command action) {
         Button button = new Button(text);
-        button.addClickListener(e -> action.run());
+        button.addClickListener(e -> action.execute());
         button.setId(id);
         return button;
     }

--- a/hummingbird-tests/test-root-context/src/test/java/com/vaadin/hummingbird/ui/SerializationTest.java
+++ b/hummingbird-tests/test-root-context/src/test/java/com/vaadin/hummingbird/ui/SerializationTest.java
@@ -1,0 +1,69 @@
+package com.vaadin.hummingbird.ui;
+/*
+ * Copyright 2000-2016 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+
+import org.junit.Test;
+
+import com.vaadin.hummingbird.router.Location;
+import com.vaadin.hummingbird.router.LocationChangeEvent;
+import com.vaadin.hummingbird.router.Router;
+import com.vaadin.hummingbird.router.View;
+import com.vaadin.hummingbird.uitest.servlet.ViewClassLocator;
+import com.vaadin.ui.UI;
+
+public class SerializationTest {
+
+    @Test
+    public void testViewsSerializable() throws Exception {
+        UI ui = new UI();
+        UI.setCurrent(ui);
+        try {
+            Collection<Class<? extends View>> viewClasses = new ViewClassLocator(
+                    getClass().getClassLoader()).getAllViewClasses();
+            for (Class<? extends View> viewClass : viewClasses) {
+                View view = viewClass.newInstance();
+                view.onLocationChange(new LocationChangeEvent(new Router(), ui,
+                        new Location(""), Collections.emptyList(),
+                        Collections.emptyMap()));
+                serializeDeserialize(view);
+            }
+        } finally {
+            UI.setCurrent(null);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T extends Serializable> T serializeDeserialize(T t)
+            throws IOException, ClassNotFoundException {
+        ByteArrayOutputStream bs = new ByteArrayOutputStream();
+        ObjectOutputStream out = new ObjectOutputStream(bs);
+        out.writeObject(t);
+        byte[] data = bs.toByteArray();
+        ObjectInputStream in = new ObjectInputStream(
+                new ByteArrayInputStream(data));
+
+        return (T) in.readObject();
+    }
+}

